edge/consumer.go

```diff
@@ -0,0 +1,69 @@
+package edge
+
+type Consumer struct {
+	edge Edge
+	r    Receiver
+}
+
+func NewConsumer(edge Edge, r Receiver) *Consumer {
+	return &Consumer{
+		edge: edge,
+		r:    r,
+	}
+}
+
+func (ec *Consumer) Run() error {
+	for msg, ok := ec.edge.Next(); ok; msg, ok = ec.edge.Next() {
+		switch typ := msg.Type(); typ {
+		case BeginBatch:
```

Comment:
I like that idea. I am too concerned that this will be less performant. For the most part I have ignored those questions and just been focusing on the API design first. Then we can come back through and fix slow spots. We will need to develop good benchmarks to test.